### DISCUSSION TITLE
The mkdocs-multirepo-plugin, dynamic generation of reference files, and git hooks

### DISCUSSION THREAD

**Discussion by kangshung** [mkdocstrings/mkdocstrings#544] [Q&A] [answered]
I use [gen-files](https://github.com/oprypin/mkdocs-gen-files) plugin with the [multirepo-plugin](https://github.com/jdoiro3/mkdocs-multirepo-plugin).

Because the latter requires me to keep all the files on the branch in the repo (it simply git clones the content of the branches, and connects multiple mkdocs configs into one), I cannot dynamically generate mkdocstrings files.

Instead, I always need to regenerate the mkdocstrings reference content with `:::`, and push these changes along with the changes in the codebase.

The question is if it's possible to automate the generation of these reference files.

I need to:
- delete reference directories that no longer exist,
- add new ones whenever new code directories with docstrings appear,
- update the mkdocs.yaml with the new structure that matches the new reference directory tree,
- and combine it with the pre-commit / pre-push git hooks.

Do you have any working examples with [pre-commit](https://github.com/pre-commit/pre-commit) or [git hooks](https://git-scm.com/docs/githooks)?

**Comment by pawamoy** [comment, marked as answer]
The script used by mkdocs-gen-files to generate the reference pages on the fly can also be executed normally, in which case it will write the pages into your docs folder. So you'd just need to run it from time to time. A small update is needed to remove directories that no longer exist:

```diff
"""Generate the code reference pages and navigation."""

+import shutil
from pathlib import Path

import mkdocs_gen_files

nav = mkdocs_gen_files.Nav()

+shutil.rmtree("docs/reference")
for path in sorted(Path("src").rglob("*.py")):
    module_path = path.relative_to("src").with_suffix("")
    doc_path = path.relative_to("src").with_suffix(".md")
    full_doc_path = Path("reference", doc_path)

    parts = tuple(module_path.parts)

    if parts[-1] == "__init__":
        parts = parts[:-1]
        doc_path = doc_path.with_name("index.md")
        full_doc_path = full_doc_path.with_name("index.md")
    elif parts[-1] == "__main__":
        continue

    nav[parts] = doc_path.as_posix()

    with mkdocs_gen_files.open(full_doc_path, "w") as fd:
        ident = ".".join(parts)
        fd.write(f"::: {ident}")

    mkdocs_gen_files.set_edit_path(full_doc_path, path)

with mkdocs_gen_files.open("reference/SUMMARY.md", "w") as nav_file:
    nav_file.writelines(nav.build_literate_nav())
```

Now just run it with `python scripts/gen_ref_nav.py` (or whatever location it's at).

If you also use mkdocs-literate-nav, then the navigation is automatically updated, no further changes are required.

**Comment by dbugajny** [comment]
@pawamoy, do you know how to combine this script with pre-commit hooks in the right way? First of all I would ask if combining a script which creates and deletes files with pre-commit is appropriate.

**Reply by pawamoy** [reply]
I'm not sure, I don't use the pre-commit tool, and rarely find pre-commit hooks useful :confused:
